Feature-initial

RRusakou opened the pull request.

RRusakou committed:
Initial commit

RRusakou committed:
Created main contract

RRusakou committed:
Create tests for MLM_System.sol

kuriloalexey commented on test/MLMSystemTest.js:
```diff
@@ -0,0 +1,159 @@
+  });
+
+  it("contract should be correctly deployed", async function () {
+    expect(mlmSystem.address).to.be.properAddress;
```
почему юзаешь именно `properAddress`?

RRusakou replied:
Не совсем правильно написал описание к тесту. properAddress вроде как проверяет правильный ли общий вид адреса и проверяет не равен ли он нулевому. Других методов на проверку правильности адреса я не нашел

kuriloalexey commented on test/MLMSystemTest.js:
```diff
@@ -0,0 +1,159 @@
+  });
+
+  it("when the user has registered without a referral, his balance should = 0", async function () {
+    await mlmSystem.connect(acc1)["signUp()"]();
+    balance = await mlmSystem.getBalance(acc1.address);
+    expect(balance == 0);
```
нужно переписать тесты все

RRusakou replied:
Прям все-все? Тогда не совсем понимаю что не так во всех тестах

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.8;
```
версия

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+    mapping(address => address[]) private directPartners;
+    mapping(address => uint256) private userBalance;
+    mapping(address => bool) private isSignedUp;
+    uint8 constant feePercents = 5;
```
константы большими буквами

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+    //Returns the level of the user, depending on his balance
+    function getLevel(address _member) public view returns (uint8) {
+        uint currentBalance = getBalance(_member);
+        if (currentBalance >= 5e15 && currentBalance < 1e16) {
+            return 1;
+        } else if (currentBalance >= 1e16 && currentBalance < 2e16) {
+            return 2;
+        } else if (currentBalance >= 2e16 && currentBalance < 5e16) {
+            return 3;
+        } else if (currentBalance >= 5e16 && currentBalance < 1e17) {
+            return 4;
+        } else if (currentBalance >= 1e17 && currentBalance < 2e17) {
+            return 5;
+        } else if (currentBalance >= 2e17 && currentBalance < 5e17) {
+            return 6;
+        } else if (currentBalance >= 5e17 && currentBalance < 1e18) {
+            return 7;
+        } else if (currentBalance >= 1e18 && currentBalance < 2e18) {
+            return 8;
+        } else if (currentBalance >= 2e18 && currentBalance < 5e18) {
+            return 9;
+        } else if (currentBalance >= 5e18) {
+            return 10;
+        }
```
сделать динамически

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+
+    //Returns the level of the user, depending on his balance
+    function getLevel(address _member) public view returns (uint8) {
+        uint currentBalance = getBalance(_member);
```
1 ether

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+        returns (uint256 comission)
+    {
+        require(level > 0 && level <= 10, "Wrong level");
+        if (level == 1) {
+            return 10;
+        } else if (level == 2) {
+            return 7;
+        } else if (level == 3) {
+            return 5;
+        } else if (level == 4) {
+            return 2;
+        } else {
+            return 1;
+        }
```
динамически

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,159 @@
+        isSignedUp[msg.sender] = true;
+    }
+
+    // Function to check the existence of the specified referrer
+    function isReferrerExist(address _referrer) private view returns (bool) {
+        return isSignedUp[_referrer];
+    }
```
Order of Functions[](https://docs.soliditylang.org/en/develop/style-guide.html#order-of-functions)

kuriloalexey commented on test/MLMSystemTest.js:
```diff
@@ -0,0 +1,159 @@
+  });
+
+  it("contract should be correctly deployed", async function () {
+    expect(mlmSystem.address).to.be.properAddress;
```
почему `properAddress`?

kuriloalexey commented on test/MLMSystemTest.js:
```diff
@@ -0,0 +1,159 @@
+  });
+
+  it("when the user has registered without a referral, his balance should = 0", async function () {
+    await mlmSystem.connect(acc1)["signUp()"]();
+    balance = await mlmSystem.getBalance(acc1.address);
+    expect(balance == 0);
```
нужно будет переписать тесты

kuriloalexey commented on package.json:
```diff
@@ -0,0 +1,253 @@
+    "test": "test"
+  },
+  "dependencies": {
+    "abort-controller": "^3.0.0",
+    "abstract-leveldown": "^6.3.0",
+    "adm-zip": "^0.4.16",
+    "aes-js": "^3.0.0",
+    "agent-base": "^6.0.2",
+    "aggregate-error": "^3.1.0",
+    "ansi-colors": "^4.1.3",
+    "ansi-escapes": "^4.3.2",
+    "ansi-regex": "^5.0.1",
+    "ansi-styles": "^3.2.1",
+    "antlr4ts": "^0.5.0-alpha.4",
+    "anymatch": "^3.1.2",
+    "argparse": "^2.0.1",
+    "assertion-error": "^1.1.0",
+    "async": "^2.6.4",
+    "async-eventemitter": "^0.2.4",
+    "balanced-match": "^1.0.2",
+    "base-x": "^3.0.9",
+    "base64-js": "^1.5.1",
+    "bech32": "^1.1.4",
+    "binary-extensions": "^2.2.0",
+    "blakejs": "^1.2.1",
+    "bn.js": "^5.2.1",
+    "brace-expansion": "^1.1.11",
+    "braces": "^3.0.2",
+    "brorand": "^1.1.0",
+    "browser-stdout": "^1.3.1",
+    "browserify-aes": "^1.2.0",
+    "bs58": "^4.0.1",
+    "bs58check": "^2.1.2",
+    "buffer": "^5.7.1",
+    "buffer-from": "^1.1.2",
+    "buffer-xor": "^2.0.2",
```
по-моему слишком много зависимостей

kuriloalexey commented:
название ветки

RRusakou committed:
Rewritten contract and test

correct sequence of functions,
solidity version changed,
rewritten check levels and commissions,
removed unnecessary dependencies,
rewritten tests

kuriloalexey commented on test/MLMSystemTest.js:
```diff
@@ -0,0 +1,160 @@
+  });
+
+  it("when the user has registered without a referral, his balance should = 0", async function () {
+    await mlmSystem.connect(acc1)["signUp()"]();
```
почему ты испольузешь через [`"signUp()"]`?

RRusakou replied:
У меня есть перегруженные функции. Регистрация для юзера без реферера и с реферером. Их вызов осуществляется таким образом...

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,155 @@
+    function findComission(uint8 level) private view returns (uint256) {
+        require(level > 0 && level <= 10, "Wrong level");
+
+        if (level < 5) {
+            return percents[level - 1];
+        } else {
+            return percents[percents.length - 1];
+        }
```
интересная логика
а если потом комиссия измениться?
 то это придется полностью переписывать

RRusakou committed:
Updated comission function

RRusakou committed:
Changed the way commissions are distributed. Rewritten tests accordingly

Functions for distributing commissions to referrs were rewritten. Changed commission transfer from wallets to internal balance.
Tests have been rewritten for the new feature

kuriloalexey commented:
название ветки

kuriloalexey commented on contracts/MLMSystem.sol:
```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.12;
+
```
Order of Functions

kuriloalexey replied:
https://docs.soliditylang.org/en/develop/style-guide.html#order-of-functions

RRusakou committed:
Order of functions

Changed order of functions.

kuriloalexey commented:
название ветки
